README.md
# Multiple Domain

[![Build Status](https://travis-ci.com/straube/multiple-domain.svg?branch=master)](https://travis-ci.com/straube/multiple-domain)

Multiple Domain allows you having more than one domain in a single WordPress installation. This plugin doesn't support
more than one theme or advanced customizations for each domain. It's only intended to enable constant navigation under
many domains. For a more complex setup, there is
[WordPress Multisite (MU)](https://codex.wordpress.org/Create_A_Network).

When there is more than one domain set in your host, all links and resources will point to the default domain. This is
the default WordPress behavior. With Multiple Domain installed and properly configured, it'll update all link on the
fly. This way, the user navigation will be end-to-end under the same domain.

You can also set an optional base URL. If you want only a set of URL's available under a given domain, you can use this
restriction.

Additionally, a language can be set for each domain. The language will be used to add `<link>` tags with `hreflang`
attribute to document head. This is for SEO purposes.

Check the plugin page at WordPress.org: https://wordpress.org/plugins/multiple-domain/

## Installation

Follow the steps below to install the plugin:

1. Upload the plugin files to the `/wp-content/plugins/multiple-domain` directory, or install the plugin through the
    WordPress plugins screen directly.
2. Activate the plugin through the 'Plugins' screen in WordPress.
3. Use the Settings -> General screen to configure your additional domains.

## Frequently Asked Questions

**Does this plugin set extra domains within my host?**

No. You have to set additional domains, DNS, and everything else to use this plugin.

**Can I have a different theme/content/plugins for each domain?**

Nope. If you want a complex set up like this, you may be interested in WordPress Multisite. It's delivered with every
WordPress installation since 3.0, you can find more info here: https://codex.wordpress.org/Create_A_Network.

**There is a way to add domain based logic to my themes?**

Absolutely. You can use the `MULTIPLE_DOMAIN_DOMAIN` and `MULTIPLE_DOMAIN_ORIGINAL_DOMAIN` constants to get the current
and original domains. Just notice that since the value of the first one is checked against plugin settings, it may not
reflect the actual domain in `HTTP_HOST` element from `$_SERVER` or user's browser. They also may include the host port
when it's different than 80 (default HTTP port) or 443 (default HTTPS port).

**Notice**: in prior versions these constants were wrongly prefixed with `MULTPLE_`, missing the "I". The old constants
are now deprecated. They still available for backcompat but will be removed in future releases.

**Can I create a custom access restriction logic for each domain?**

Yes. You can use the `multiple_domain_redirect` action to do that. Please check
https://github.com/straube/multiple-domain/issues/2 for an example on how to do that.

**Can I get the language associated with the current domain?**

Yes. You can use the `MULTIPLE_DOMAIN_DOMAIN_LANG` constant to get the language associated with the current domain. Keep
in mind the value in this constant doesn't necessarily reflect the actual user language or locale. This is just the
language set in the plugin config. Also notice the language may be `null`.

**Notice**: in prior versions these constants were wrongly prefixed with `MULTPLE_`, missing the "I". The old constants
are now deprecated. They still available for backcompat but will be removed in future releases.

**Can I show the current domain in the content of posts or pages?**

Yes. There is a shortcode available for that. Just add `[multiple_domain]` to the post/page and it'll be replaced by
the current domain when viewing the content. You can write things like "Welcome to [multiple_domain]!", which would be
rendered as "Welcome to mydomain.com!".

**What domains should I add to the plugin setup?**

Any domain you're site is served from must be added to the plugin configuration. Even `www` variations and the original
domain where your WordPress was installed in must be added. You'll probably see some unexpected output when accessing
the site from a non-mapped domain.

**Can I disable `hreflang` tags output even for the original domain?**

Yes. You may notice that even if you don't set a language for any domain, you still get a default `hreflang` tag in
your page head. To disable this behavior, follow the instructions from
https://github.com/straube/multiple-domain/issues/51.

**I locked myself out, and what am I doing now?**

Under certain circumstances, in the case of a wrong configuration, you may not be able to log in to the admin area
and your page will be redirected. In this case, there are two ways to solve this.

1. Delete the plugin directory `wp-content/plugins/multiple-domain`. You should be able to do that from the hosting
    panel, from an FTP client, or via SSH. The downside of this technique is that it won’t be possible to install the
    plugin again since the configuration will still be in the database.
2. Remove the plugin configuration from the database using the following SQL query `DELETE FROM {YOUR-PREFIX}_options
    WHERE option_name LIKE 'multiple-domain-%'`; (Remember to replace the prefix from your own table name). This can be
    done from the hosting panel when PHPMyAdmin is available or using a MySQL client.

## Changelog

Refer to [CHANGELOG.md](CHANGELOG.md) for release notes and changes.

## Meta

Contributors: GustavoStraube, cyberaleks, jffaria  
Tags: multiple, domains, redirect  
Requires at least: 4.0  
Tested up to: 5.2.4  
Stable tag: trunk  
License: GPLv2 or later  
License URI: http://www.gnu.org/licenses/gpl-2.0.html  

## Contributing

Want to help? Please check our [contribution guide](CONTRIBUTING.md).


CHANGELOG.md
# Release Notes

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
Since `0.7.1` this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.6] - 2020-08-02

### Fixed

* Fix URI generated for canonical tag. ([#76])

## [1.0.5] - 2020-02-03

### Fixed

* Fixed issue with system routes when a base path is defined.

## [1.0.4] - 2020-01-25

### Fixed

* Fixed assertions in admin views.

## [1.0.3] - 2020-01-25

### Security

* Fixed XSS vulnerability in canonical/alternate tags. ([#71])

## [1.0.2] - 2019-10-28

### Added

* Low memory option. ([#45])

### Deprecated

* Constants starting with `MULTPLE_` are now deprecated. They have a matching `MULTIPLE_` prefixed constant.

### Fixed

* Fixed constants starting with `MULTPLE_`, changed to `MULTIPLE`.

## [1.0.1] - 2019-10-25

### Fixed

* Fixed issue with regex used in domain replacement. ([#59])

## [1.0.0] - 2019-10-21

### Added

* Locked out instructions to readme file. ([#61])
* API to programmatically change the domains list. ([#56])

### Fixed

* Don't add canonical link if settings are `false`. ([#58])

## [0.11.2] - 2019-04-19

### Added

* FAQ about removal of `hreflang` tags. ([#51])

### Fixed

* Bug in domain replacement when it contains a slash (the regex delimiter).
* Fixed issue in the domain replacement regex. ([#55])

## [0.11.1] - 2019-04-10

### Fixed

* Improved URI validation when there is a domain's base restriction. ([#50])

## [0.11.0] - 2019-04-09

### Added

* Add [CHANGELOG.md] file. ([#44])
* Added option to enable canonical tags. ([#42])
* Added `%%multiple_domain%%` advanced variable for Yoast.

### Changed

* Moved WordPress admin features to a separate class.
* Renamed hreflang related methods.
* Inline documentation review.
* Minor refactoring.

### Fixed

* Fixed issue with domain replacement.

## [0.10.3] - 2019-03-12

### Added

* Added public method to retrieve a given (or current) domain base path: `getDomainBase($domain = null)`.

### Changed

* Minor code refactoring.

## [0.10.2] - 2019-03-11

### Added

* Added FAQ about plugin settings and domains.

### Fixed

* Fix minor notice message when loading the non-mapped original domain.

## [0.10.1] - 2019-03-11

### Fixed

* Fix bug introduced in 0.10.0 with setups where the original domain is not present in the plugin settings.

## [0.10.0] - 2019-03-08

### Changed

* Moved HTML to view files.

### Fixed

* Don't add SSL when accessing via a Tor domain name. ([#31])

## [0.9.0] - 2019-02-09

### Added

* Added a class to `<body>` tag containing the domain name (e.g. `multipled-domain-name-tld`) to allow front-end customizations.

### Fixed

* Fixed bug in backward compatibility logic.

## [0.8.7] - 2019-02-07

### Changed

* Loading Multiple Domain before other plugins to fix issue with paths.
* Refactored `initAttributes` method.

### Fixed

* Missing locales on language list (this issue was reopened and now it's fixed) ([#38])

## [0.8.6] - 2019-02-01

### Removed

* Rolling back changes introduced in [0.8.4] and [0.8.5] regarding to avoid URL changes in the WP admin. ([#39])

## [0.8.5] - 2019-01-18

### Added

* Add `[multiple_domain]` shortcode to show the current language.

### Fixed

* Fixed an issue introduced in 0.8.4 that breaks the admin URLs.
* Missing locales on language list ([#38])

## [0.8.4] - 2019-01-16

### Changed

* Using singleton pattern for main plugin class.
* Avoiding URL changes in the admin panel.

### Fixed

* Wrong host in URLs returned by the JSON API. ([#36])

## [0.8.3] - 2019-01-12

### Fixed

* `hreflang` tag error. ([#34])

## [0.8.2] - 2018-12-19

### Fixed

* Image URLs not being re-written properly via Tor. ([#32])

## [0.8.1] - 2018-10-09

### Fixed

* Undefined index when using wp-cli. ([#23])

## [0.8.0] - 2018-09-07

### Added

* Added `MULTIPLE_DOMAIN_DOMAIN_LANG` constant for theme/plugin customization. ([#20])

### Changed

* Moved `MultipleDomain` class to its own file.

### Fixed

* No 'Access-Control-Allow-Origin' header is present on the requested resource. ([#21])
* Attempt to fix #22. ([#22])

### Security

* Remove `filter_input` from plugin. ([#14])

## [0.7.1] - 2018-05-25

### Fixed

* Make the plugin compatible with PHP 5.4 again.

## [0.7] - 2018-05-24

Notice that version `0.6` wasn't tagged as a release, then `0.7` also includes its changes.

### Changed

* Code review/refactoring.

### Fixed

* Added activation hook to fix empty settings bug.
* Redirect to original domain if SSL/https ([#11]).

## [0.5] - 2018-01-04

### Added

* `http`/`https` for alternate link.

## [0.4] - 2017-11-28

### Added

* Added Reflang links to head for SEO purpose.  
e.g.
```html
<link rel="alternate" hreflang="x-default" href="https://example.com/">
<link rel="alternate" hreflang="de-DE" href="https://de.example.com/">
```

### Fixed

* Fixed resolving host name to boolean.

## [0.3] - 2016-08-21

### Added

* `MULTIPLE_DOMAIN_ORIGINAL_DOMAIN` constant to hold the original WP home domain.
* Allowing developers to create custom URL restriction logic through `multiple_domain_redirect` action.

### Changed

* Improved settings interface.

### Fixed

* Fixed bug when removing the port from current domain.

## [0.2] - 2016-07-28

### Added

* `MULTIPLE_DOMAIN_DOMAIN` constant for theme/plugin customization.

### Changed

* Improved port verification.
* And, last but not least, code refactoring.

## [0.1] - 2016-07-28

### Added

* Basic multiple domain setup.
* Option base path for each domain.


[Unreleased]: https://github.com/straube/multiple-domain/compare/v1.0.6...HEAD
[1.0.6]: https://github.com/straube/multiple-domain/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/straube/multiple-domain/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/straube/multiple-domain/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/straube/multiple-domain/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/straube/multiple-domain/compare/v1.0.1...v1.0.2

multiple-domain/multiple-domain.php
<?php

/*
Plugin Name: Multiple Domain
Plugin URI:  https://github.com/straube/multiple-domain
Description: This plugin allows you to have multiple domains in a single WordPress installation and enables custom redirects for each domain.
Version:     1.0.6
Author:      Gustavo Straube (straube.co)
Author URI:  http://straube.co
License:     GPLv2 or later
License URI: https://www.gnu.org/licenses/gpl-2.0.html
Text Domain: multiple-domain
Domain Path: /languages
*/


/*
 * Loading classes.
 */
require 'MultipleDomain.php';
require 'MultipleDomainSettings.php';


/**
 * The plugin file name.
 *
 * This is used mainly to set hooks and other features that requires the base
 * plugin file name to work properly.
 *
 * @var   string
 * @since 0.8.3
 */
define('MULTIPLE_DOMAIN_PLUGIN', __FILE__);

if (!defined('MULTIPLE_DOMAIN_LOW_MEMORY')) {

    /**
     * The low memory option.
     *
     * This option may be used where the site is throwing "allowed memory
     * exhausted" errors. It will reduce the memory usage in domain replacements
     * with the downside of a higher execution time.
     *
     * @var   bool
     * @since 1.0.2
     */
    define('MULTIPLE_DOMAIN_LOW_MEMORY', false);
}


/*
 * Register the activation method.
 */
register_activation_hook(MULTIPLE_DOMAIN_PLUGIN, [ MultipleDomain::class, 'activate' ]);


/*
 * Bootstrap...
 */
$multipleDomain = MultipleDomain::instance();
$domain = $multipleDomain->getDomain();
$originalDomain = $multipleDomain->getOriginalDomain();
$domainLang = $multipleDomain->getDomainLang();


/**
 * The current domain.
 *
 * Since this value is checked against plugin settings, it may not reflect the
 * actual domain in `HTTP_HOST` element from `$_SERVER`. It also may include
 * the host port when it's different than 80 (default HTTP port) or 443
 * (default HTTPS port).
 *
 * @var   string
 * @since 1.0.2
 */
define('MULTIPLE_DOMAIN_DOMAIN', $domain);


/**
 * The original domain set in WordPress installation.
 *
 * @var   string
 * @since 1.0.2
 */
define('MULTIPLE_DOMAIN_ORIGINAL_DOMAIN', $originalDomain);


/**
 * The current domain language.
 *
 * This value is the language associated with the current domain in the plugin
 * settings. No check is made to verifiy if it reflects the actual user
 * language or locale. Also, notice this constant may be `null` when no
 * language is set in the plugin config.
 *
 * @var   string
 * @since 1.0.2
 */
define('MULTIPLE_DOMAIN_DOMAIN_LANG', $domainLang);


/**
 * Keeping back compability with prior versions.
 *
 * This constant will be removed in a future release.
 *
 * @var   string
 * @since 0.2
 * @see   MULTIPLE_DOMAIN_DOMAIN
 * @deprecated
 */
define('MULTPLE_DOMAIN_DOMAIN', MULTIPLE_DOMAIN_DOMAIN);


/**
 * Keeping back compability with prior versions.
 *
 * This constant will be removed in a future release.
 *
 * @var   string
 * @since 0.3
 * @see   MULTIPLE_DOMAIN_ORIGINAL_DOMAIN
 * @deprecated
 */
define('MULTPLE_DOMAIN_ORIGINAL_DOMAIN', MULTIPLE_DOMAIN_ORIGINAL_DOMAIN);


/**
 * Keeping back compability with prior versions.
 *
 * This constant will be removed in a future release.
 *
 * @var   string
 * @since 0.8
 * @see   MULTIPLE_DOMAIN_DOMAIN_LANG
 * @deprecated
 */
define('MULTPLE_DOMAIN_DOMAIN_LANG', MULTIPLE_DOMAIN_DOMAIN_LANG);
